query-engine/query-engine-node-api/src/node_drivers/engine.rs

```diff
@@ -0,0 +1,165 @@
+use crate::engine::QueryEngine;
+
+use napi::bindgen_prelude::Promise as JsPromise;
+use napi::{
+    threadsafe_function::{ErrorStrategy, ThreadsafeFunction},
+    Env, JsFunction, JsObject, JsUnknown,
+};
+use napi_derive::napi;
+use serde::{Deserialize, Serialize};
+
+#[napi(object)]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ResultSet {
+    pub columns: Vec<String>,
+    pub rows: Vec<Vec<String>>,
+}
+
+// Note: Every ThreadsafeFunction<T, ?> should have an explicit `ErrorStrategy::Fatal` set, as to avoid
+// "TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received null".
+// See: https://github.com/napi-rs/napi-rs/issues/1521.
+pub(crate) struct NodejsFunctionContext {
+    // TODO: maybe env is no longer needed
+    _env: Env,
+
+    /// Execute a query given as SQL, interpolating the given parameters.
+    pub query_raw: ThreadsafeFunction<String, ErrorStrategy::Fatal>,
+
+    /// Execute a query given as SQL, interpolating the given parameters and
+    /// returning the number of affected rows.
+    pub execute_raw: ThreadsafeFunction<String, ErrorStrategy::Fatal>,
+
+    /// Return the version of the underlying database, queried directly from the
+    /// source.
+    pub version: ThreadsafeFunction<(), ErrorStrategy::Fatal>,
+}
+
+unsafe impl Send for NodejsFunctionContext {}
+unsafe impl Sync for NodejsFunctionContext {}
```

Comment:
If we stick with just `ThreadsafeFunction<T>` in `NodejsFunctionContext`, we will be able to remove those `unsafe impl`s, so let's not worry about them right now.